oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/jdbc/shardingsphere/ShardingSphereTableInstaller.java

```diff
@@ -0,0 +1,97 @@
+import org.apache.skywalking.oap.server.core.storage.StorageException;
+import org.apache.skywalking.oap.server.core.storage.model.Model;
+import org.apache.skywalking.oap.server.core.storage.model.ModelColumn;
+import org.apache.skywalking.oap.server.library.client.Client;
+import org.apache.skywalking.oap.server.library.client.jdbc.JDBCClientException;
+import org.apache.skywalking.oap.server.library.client.jdbc.hikaricp.JDBCHikariCPClient;
+import org.apache.skywalking.oap.server.library.module.ModuleManager;
+import org.apache.skywalking.oap.server.storage.plugin.jdbc.h2.H2TableInstaller;
+
+/**
+ * ShardingSphere table installer delegates all the works to
+ * the table installer (e.g. H2TableInstaller, MySQLTableInstaller)
+ * of the backed storage (e.g. H2, MySQL), and then add sharding rules.
+ */
+public class ShardingSphereTableInstaller extends H2TableInstaller {
+    private final H2TableInstaller delegatee;
+    private final Set<String> dataSources;
+    private final ModuleManager moduleManager;
+
+    public ShardingSphereTableInstaller(Client client,
+                                        ModuleManager moduleManager,
+                                        Set<String> dataSources,
+                                        H2TableInstaller delegatee) {
+        super(client, moduleManager);
+        this.dataSources = dataSources;
+        this.delegatee = delegatee;
+        this.moduleManager = moduleManager;
+    }
+
+    @Override
+    public boolean isExists(Model model) throws StorageException {
+        boolean isRuleExecuted = false;
+        boolean isTableExist = delegatee.isExists(model);
+        JDBCHikariCPClient jdbcClient = (JDBCHikariCPClient) client;
+        ConfigService configService = moduleManager.find(CoreModule.NAME).provider().getService(ConfigService.class);
+        int ttl = model.isRecord() ? configService.getRecordDataTTL() : configService.getMetricsDataTTL();
+        if (model.getSqlDBModelExtension().isShardingTable()) {
+            isRuleExecuted = ShardingRulesOperator.INSTANCE.createOrUpdateShardingRule(jdbcClient, model, this.dataSources, ttl);
+        }
+        return isTableExist && !isRuleExecuted;
+    }
+
+    @Override
+    public void start() {
+        delegatee.start();
+    }
+
+    @Override
+    public void createTable(
+        JDBCHikariCPClient client,
+        Connection connection,
+        String tableName,
+        List<ModelColumn> columns,
+        boolean additionalTable) throws JDBCClientException {
+        delegatee.createTable(client, connection, tableName, columns, additionalTable);
+    }
+
+    @Override
+    public void createTableIndexes(
+        JDBCHikariCPClient client,
+        Connection connection,
+        String tableName,
+        List<ModelColumn> columns,
+        boolean additionalTable) throws JDBCClientException {
+        delegatee.createTable(client, connection, tableName, columns, additionalTable);
```

Comment:
Should invoke `createTableIndexes`?
If you want to reuse the super method, should better test the index creation by using the `sharding-proxy`, 
when I try to reuse it before, I got many issues, such as the index name overlength, and can not create index success in this way.

Reply:
The index seems can not be covered by the e2e.

Reply:
This is what I changed:

1. We previously create table and then create table indices separately for MySQL storage, and
2. use create table with indices (`create table xxx (yyy int, index key0...)`), for MySQL ShardingSphere storage.

The method (2) applies to both MySQL and ShardingSphere, but method (1) only applies to MySQL storage, to unify the two implementations, I changed (1) to (2) as well so we don't need different table installer for these two implementations.